tests/bluetooth/bsim_bt/bsim_test_mesh/src/settings_test_backend.c

```diff
@@ -0,0 +1,218 @@
+{
+	struct line_read_ctx *valctx = (struct line_read_ctx *)cb_arg;
+	char val[ENTRY_VAL_MAX_LEN + 1];
+
+	if (fseek(fp, valctx->offset, SEEK_SET) < 0 ||
+	    fscanf(fp, val_fmt, val) != 1 ||
+	    (valctx->len / 2) != len) {
+		return 0;
+	}
+
+	uint8_t *n = (uint8_t *) data;
+
+	for (int i = 0; i < len; i++, n++) {
+		if (sscanf(&val[i * 2], "%2hhx", n) != 1) {
+			return 0;
+		};
+	}
+
+	return len;
+}
+
+static int settings_custom_load(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	if (fseek(fp, 0, SEEK_SET) < 0) {
+		return -1;
+	}
+
+	int vallen;
+	char name[ENTRY_NAME_MAX_LEN + 1];
+
+	while (fscanf(fp, len_name_fmt, &vallen, name) == 2) {
+
+		if (vallen <= 0) {
+			continue;
+		}
+
+		/* check for matching subtree */
+		if (arg->subtree != NULL && !strstr(name, arg->subtree)) {
+			continue;
+		}
+
+		LOG_INF("loading entry: %s", name);
+
+		struct line_read_ctx valctx;
+
+		valctx.len = vallen;
+		valctx.offset = ftell(fp);
+		int err = settings_call_set_handler(name, vallen / 2, settings_line_read_cb,
+						    &valctx, arg);
+
+		if (err < 0) {
+			return err;
+		}
+	};
+
+	return 0;
+}
+
+/* Entries are saved to optimize redability of the settings file for test development and
+ * debugging purposes. Format:
+ * <length-hex-str><sapce><entry-key><space><entry-value-hex-str>
+ */
+static int settings_custom_save(struct settings_store *cs, const char *name,
+				const char *value, size_t val_len)
+{
+	if (strlen(name) > ENTRY_NAME_MAX_LEN || val_len > SETTINGS_MAX_VAL_LEN) {
+		return -1;
+	}
+
+	if (entry_check_and_seek(name) == 0) {
+		char bufvname[ENTRY_NAME_MAX_LEN + ENTRY_LEN_SIZE + 3];
+
+		snprintk(bufvname, sizeof(bufvname), "%04x %s ", val_len * 2, name);
+		if (fputs(bufvname, fp) < 0) {
+			return -1;
+		}
+	}
+
+	char bufval[ENTRY_VAL_MAX_LEN + 2];
+	size_t valcnt = 0;
+
+	while (valcnt < (val_len * 2)) {
+		valcnt += snprintk(&bufval[valcnt], 3, "%02x",
+				   (uint8_t)value[valcnt / 2]);
+	};
+
+	/* helps in making settings file redable */
+	bufval[valcnt++] = '\n';
+	bufval[valcnt] = 0;
+
+	if (fputs(bufval, fp) < 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+/* custom backend interface */
+static struct settings_store_itf settings_custom_itf = {
+	.csi_load = settings_custom_load,
+	.csi_save = settings_custom_save,
+};
+
+/* custom backend node */
+static struct settings_store settings_custom_store = {
+	.cs_itf = &settings_custom_itf
+};
+
+int settings_backend_init(void)
+{
+	if (fseek(fp, 0, SEEK_SET) < 0) {
+		LOG_ERR("Call settings_test_backend_preinit() before starting test");
+		return -1;
+	}
+
+	sprintf(len_name_fmt, "%%04x %%%ds", ENTRY_NAME_MAX_LEN);
+	sprintf(val_fmt, " %%%ds", ENTRY_VAL_MAX_LEN);
+	sprintf(sk_fmt, "%%%ds", READ_LEN_MAX);
+
+	/* register custom backend */
+	settings_dst_register(&settings_custom_store);
+	settings_src_register(&settings_custom_store);
+	return 0;
+}
+
+int settings_test_backend_preinit(enum test_file_state fmode)
```

Comment:
Is there any reason to keep this outside the load and save calls? Putting the responsibility of opening and closing the files on the test implementation introduces a source of potential bug in every place where it's used. Couldn't we just open for reading on load and for writing on save? Both start with a SEEK_SET anyway.

If necessary, create a `settings_clear()` function that just wipes the contents

Reply:
Yes.  Especially for file path creation.  If this is integrated in "settings_backend_init()" for the purpose of opening file on each load/save call, then test code does not have a way to clear the file before the stack can load. Since "settings_backend_init()" and subsequent load happens monotonically as a result of "bt_mesh_test_setup()" calls. This structure helps maintain this separation and also ensures that the test procedures do exactly what they are supposed to do (start with a clean file, and expect to load existing file).

Reply:
The file name does not need to sit in a static variable though. It also does not need to get prepended with the cwd, as fopen paths are relative to cwd by default, but if you prefer, you could wrap fopen in a function that hardcodes the file path and accepts the access mode and returns `FILE *`. 

If preinit with the FILE_CLEAR argument is equivalent to "wipe file", why not just create a function like 
```c
void test_data_clear(void)
{
    FILE *f;
    
    f = fopen(SETTINGS_FILE, "w");
    if (f) {
        fclose(f);
    }
}
```
?

I appreciate the value in mimicking the target behavior, but if it's forcing us to create bug generators above the API (in the form of missing fclose calls), it's just not a strong enough argument for me. I don't see any concrete, unavoidable arguments for creating an API with a built-in footgun, so I don't think we should do it.